ACL Management, File Expiry, Consistent Credentials and API Management

begoldsm opened the pull request:
Fix creds to use the generic msrest.authentication class.
Fix ADLDownloader to do validation of the source path before creating
local files.

Update history

Add new methods and tests

Adds set_expiry and ACL management functions and tests. Also adds
support for API versions and fixes up the tests to use the new
credential logic.

lmazuel commented on azure/datalake/store/core.py:
```diff
@@ -192,6 +196,152 @@ def chmod(self, path, mod):
+            parms['expireTime'] = int(expire_time)
+
+        self.azure.call('SETEXPIRY', path.as_posix(), is_extended=True, **parms)
+    def set_acl(self, path, acl_spec):
```
New line (to separate methods)

lmazuel commented on azure/datalake/store/core.py:
```diff
@@ -192,6 +196,152 @@ def chmod(self, path, mod):
+                4. Absolute: ExpireTime is an integer in milliseconds, as a Unix timestamp relative to 1/1/1970 00:00:00.
+        """
+        parms = {}
+        valid_expire_types = ['neverexpire', 'relativetonow', 'relativetocreationdate', 'absolute']
```
Optional: there is enum in Python

lmazuel commented on azure/datalake/store/lib.py:
```diff
@@ -25,6 +25,7 @@
 import requests
 import requests.exceptions
 
+from msrest.authentication import (Authentication)
```
Parenthesis useless

lmazuel commented on azure/datalake/store/lib.py:
```diff
@@ -141,23 +143,43 @@ def auth(tenant_id=None, username=None,
 
+    return DataLakeCredential(out)
+
+class DataLakeCredential(Authentication):
```
We need to discuss that, I wasn't able to understand your issue in the email thread. Why do you need here something that is supposed to be generic? Or you have some ADL behavior that requires a specific class?

begoldsm replied:
@lmazuel we briefly synced separately on this. I will work with you a bit later to see if we can further generalize the Auth so that I can enable the ability to use any signed_session() enabled credential.

lmazuel commented on azure/datalake/store/lib.py:
```diff
@@ -177,21 +199,36 @@ class DatalakeRESTInterface:
+
+        # There is a case where the user can opt to exclude an API version, in which case
+        # the service itself decides on the API version to use (it's default).
+        if api_version:
```
self.api_version = api_version or None

lmazuel replied:
the "or" operator returns the first value than can be considered true using "bool" method. And there is no such element, returns the latest element.
1 or 2 => 1
0 or 2 => 2
0 or [] or {} or "" or 15 => 15

brettcannon commented on azure/datalake/store/core.py:
```diff
@@ -192,6 +197,154 @@ def chmod(self, path, mod):
+        parms = {}
+        valid_expire_types = [x.value for x in ExpiryOptionType]
+        value_to_use = [x for x in valid_expire_types if x.lower() == expiry_option.lower()]
+        if not value_to_use or len(value_to_use) != 1:
```
Could this be simplified to `len(value_to_use) > 1`?

brettcannon commented on azure/datalake/store/core.py:
```diff
@@ -192,6 +197,154 @@ def chmod(self, path, mod):
+        """
+        path = AzureDLPath(path).trim()
+
+        return self.azure.call('MSGETACLSTATUS', path.as_posix())['AclStatus']
```
I count 3 uses of this pattern (5 if you count the two below where `parm` wasn't needed). It might be worth factoring this out to a helper function to cut down on repetitiveness, e.g.:
```python
def _acl_call(self, action, path, **kwargs):
    path = AzureDLPath(path).trim()
    self.azure.call(action, path.as_posix(), **kwargs)
```

Then all 5 instances become something like `self._acl_call('REMOVEACLENTRIES', path, aclSpec=acl_spec)`.

brettcannon commented on azure/datalake/store/lib.py:
```diff
@@ -214,9 +248,9 @@ def session(self):
-            self.token = refresh_token(self.token)
-            self.head = {'Authorization': 'Bearer ' + self.token['access']}
+        cur_session = self.token.signed_session()
+        if not self.head or not 'Authorization' in self.head or self.head['Authorization'] != cur_session.headers['Authorization']:
```
Could possibly simplify this to `if not self.head or self.head.get('Authorization') != cur_session.headers['Authorization']`.

brettcannon commented on azure/datalake/store/transfer.py:
```diff
@@ -447,18 +447,28 @@ def run(self, nthreads=None, monitor=True, before_start=None):
         if monitor:
             self.monitor()
+            has_errors = False
+            error_list = list()
```
Can also use `error_list = []` (unless you're overriding `list()` in tests or something 😉 ).

brettcannon commented on tests/settings.py:
```diff
@@ -19,12 +19,14 @@
-        resource="https://management.core.windows.net/",
-        tenant=TENANT_ID, expiresIn=3600)
+    TOKEN = DataLakeCredential(
+        dict(
```
Why the call to dict instead of using a dict literal?

begoldsm replied:
Probably because I don't know what a dict literal is :) Also to just minimize the amount of changes, since this was already initializing the dictionary using dict()

brettcannon replied:
So a dict literal for this would be:
```python
{
  'access': str(base64.b64encode(os.urandom(1420)),
  'refresh': str(base64.b64encode(os.urandom(718))),
  ...
}
```

brettcannon commented on tests/test_core.py:
```diff
@@ -633,22 +635,119 @@ def test_chmod(azure):
+    acl_to_add = 'user:{}:rwx'.format(user_id)
+    acl_to_modify = 'user:{}:-w-'.format(user_id)
+    acl_to_remove = 'user:{}'.format(user_id)
+    with azure_teardown(azure):
```
Is `azure` a pytest fixture? If so you should be able to have the fixture yield the value for `azure` and do proper teardown in the fixture itself.

begoldsm replied:
It is but I am not quite sure I understand the meaning, where would the teardown happen? inside of the def azure()?

begoldsm replied:
@brettcannon given that this is in the test and is being used with this pattern for all other tests as well, I am hesitant to make changes to how the tests run. This is the "don't fix what isn't broken" strategy :)

begoldsm commented on azure/datalake/store/core.py:
```diff
@@ -192,6 +197,154 @@ def chmod(self, path, mod):
+        """
+        parms = {}
+        valid_expire_types = [x.value for x in ExpiryOptionType]
+        value_to_use = [x for x in valid_expire_types if x.lower() == expiry_option.lower()]
```
@brettcannon is there something like a "select" that I can do so it comes back as just one element? It will always be just one element so I would like to be able to just get that.

brettcannon replied:
There are two options. One is a loop with a break on the first match. The other is `next(filter(lambda x: x.lower() == expiry_option.lower(), valid_expire_types))`.

begoldsm commented:
@brettcannon @lmazuel I believe I have addressed all of the comments (or commented on why I didn't make the change). Please let me know if there is anything else otherwise I look forward to your sign off! If I don't hear back today I am going to merge since we need to get this published and available for CLI consumption.
